### PR TITLE
Update RRD IDL to PPX

### PR DESCRIPTION
**Pull request by jonludlam**
Mostly due to @minishrink 

Also required:

https://github.com/xapi-project/ocaml-rrdd-plugin/pull/25
https://github.com/xapi-project/xapi-storage-script/pull/64
https://github.com/xapi-project/xen-api/pull/3582
https://github.com/xapi-project/xcp-rrdd/pull/81
https://github.com/xapi-project/xenopsd/pull/506

and requires ocaml-rpc > v4.0.0

### CODE REVIEW

File: rrd/rrd_interface.ml [hunk 1]
```diff
@@ -32,92 +35,375 @@ let set_sockets_dir x =
...
+
+exception Rrdd_error of rrd_errors
+
+(** Error handler *)
+module RrdErrHandler = Error.Make(struct
+    type t = rrd_errors
+    let t  = rrd_errors
+    let internal_error_of e = Some (Rrdd_internal_error (Printexc.to_string e))
+  end)
+let rrd_err = Error.{ def     = rrd_errors
```
Comment (mseri) [inline rrd/rrd_interface.ml]: Isn't this just `let rrd_err = RrdErrHandler.error`?

File: rrd/rrd_interface.ml [hunk 2]
```diff
@@ -32,92 +35,375 @@ let set_sockets_dir x =
...
+      (unit_p @-> returning unit_p rrd_err)
+  end
+
+  module Deprecated = struct
 
-module Deprecated = struct
-	(* Could change timescale to sum type, e.g. Slow | Fast.*)
-	external load_rrd : uuid:string -> timescale:int ->
-		master_address:string option -> unit = ""
+    let load_rrd = (* TODO: there can only be one *)
```
Comment (mseri) [inline rrd/rrd_interface.ml]: What is the TODO about?

File: rrd/rrd_interface.ml [hunk 3]
```diff
@@ -32,92 +35,375 @@ let set_sockets_dir x =
...
+    let t  = rrd_errors
+    let internal_error_of e = Some (Rrdd_internal_error (Printexc.to_string e))
+  end)
+let rrd_err = Error.{ def     = rrd_errors
+                    ; raiser  = (fun e -> raise (Rrdd_error e))
+                    ; matcher = (function
+                      | Rrdd_error e -> Some e
+                      | e            -> Some (Rrdd_internal_error (Printexc.to_string e)))
+                    }
+
```
Comment (mseri) [inline rrd/rrd_interface.ml]: I think you need to register a printer as we did for the other modules, to make sure that we do not loose the content of the errors:
```ocaml
let () = (* register printer *)
  let sprintf = Printf.sprintf in
  let string_of_error e =
    Rpcmarshal.marshal rrd_err.Rpc.Types.ty e |> Rpc.to_string in
  let printer = function
    | Gpumon_error e ->
        Some (sprintf "Rrd_interface.Rrdd_error(%s)" (string_of_error e))
    | _ -> None in
Printexc.register_printer printer
```

File: rrd/rrd_interface.ml (hunk 2 above)
Comment (minishrink) [reply rrd/rrd_interface.ml]: Ah, that can be removed now. I think this was an issue with namespaces but it's been resolved now.

**Pull request closed without merging**